### PR TITLE
ACR cleanups and fixes

### LINKED ISSUE #32810: [BUG] ACR digest and tag issues
1. Tags API does not accept digests, but we use it here

https://github.com/Azure/azure-sdk-for-java/blob/94e5e4ab4dd2035352c9e2e36cfa13c804795bec/sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/RegistryArtifact.java#L102

https://limolkovaacr.azurecr.io/acr/v1/library%2Fhello-world/_tags/unknown%253Adigest?api-version=2021-07-01

2. we need to url-encode digests:

`https://REDACTED.azurecr.io/acr/v1/library%2Fhello-world/_manifests/unknown:digest?api-version=2021-07-01`

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #32810 and few other improvements including better test coverage
Turns out the problem is in tests that sends invalid digest  (not hex) which prevents the service from routing this request and returning text/plan response without error info

### CODE REVIEW

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java [hunk 1]
```diff
@@ -29,136 +28,131 @@
...
+            return httpResponse;
+        }
+
+        return attemptRedirectSync(context, next, httpResponse, 1, new HashSet<>());
+    }
+
+    private static HttpResponse mapResponse(HttpResponse oldResponse, HttpResponse newResponse) {
+        String digest = getDigestFromHeader(oldResponse.getHeaders());
+        if (digest != null) {
+            newResponse.getHeaders().set(DOCKER_DIGEST_HEADER_NAME, digest);
```
Comment (alzimmermsft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java:62]: Is `DOCKER_DIGEST_HEADER_NAME` an `HttpHeaderName`?

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java [hunk 2]
```diff
@@ -29,136 +28,131 @@
  */
 public final class ContainerRegistryRedirectPolicy implements HttpPipelinePolicy {
```
Comment (alzimmermsft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java:29]: Long term should this transition to using the azure-core redirect policy?

**Review by alzimmermsft — APPROVED** [review]

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java (hunk 1 above)
Comment (lmolkova) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java:62]: yep!

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java (hunk 2 above)
Comment (lmolkova) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java:29]: Yes, my guess that we can reuse it directly, but I need to investigate - https://github.com/Azure/azure-sdk-for-java/issues/33183

**Pull request merged**